### PR TITLE
add faster rcnn coco dataset

### PR DESCRIPTION
**Pull request by ijkguo**
- coco dataset: option to use crowd instances; roi_align sample ratio; nms threshold; bilinear interpolation; box_decode clip
- sampler: replaced with native op
- box_clip: replaced with native op; introduced mxnet 20180713 dependency for shape_array
- multi batch inference: almost correct, except that box_clip need actual image size when padded
- resnetv1b models: new test for use_global_stats in modelzoo require reinstall pip packages

### CODE REVIEW

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py [hunk 1]
```diff
@@ -133,105 +156,90 @@ def hybrid_forward(self, F, x, gt_box=None):
...
-
-            results.append(per_result)
-        result = F.concat(*results, dim=0).expand_dims(0)
-        if self.nms_thresh > 0 and self.nms_thresh < 1:
-            result = F.contrib.box_nms(
-                result, overlap_thresh=self.nms_thresh, topk=self.nms_topk,
-                id_index=0, score_index=1, coord_start=2)
-            if self.post_nms > 0:
-                result = result.slice_axis(axis=1, begin=0, end=self.post_nms).squeeze(axis=0)
+        for rpn_box, cls_id, score, box_pred in zip(rpn_boxes, cls_ids, scores, box_preds):
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: concat class dim to batch dim, do NMS might be faster than loop.

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py [hunk 2]
```diff
@@ -133,105 +156,90 @@ def hybrid_forward(self, F, x, gt_box=None):
...
+            # box_pred (C, N, 4) rpn_box (1, N, 4) -> bbox (C, N, 4)
+            bbox = self.box_decoder(box_pred, self.box_to_center(rpn_box))
+            # res (C, N, 6)
+            res = F.concat(*[cls_id, score, bbox], dim=-1)
+            # res (C, self.nms_topk, 6)
+            res = F.contrib.box_nms(
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: valid_thresh is faster

File: gluoncv/model_zoo/faster_rcnn/rcnn_target.py [hunk 3]
```diff
@@ -4,70 +4,99 @@
...
+                # match to argmax iou
+                ious_max = ious.max(axis=-1)
+                ious_argmax = ious.argmax(axis=-1)
+                # init with -1, which are neg samples
+                mask = F.ones_like(ious_max) * -1
+                # mark positive samples with 1
+                pos_mask = ious_max >= self._pos_iou_thresh
+                mask = F.where(pos_mask, F.ones_like(mask), mask)
+
+                # shuffle mask
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/rcnn_target.py]: skeptical about the random performance here

File: gluoncv/model_zoo/rcnn/rcnn.py [hunk 4]
```diff
@@ -100,7 +101,7 @@ def collect_train_params(self, select=None):
             return self.collect_params(self.train_patterns)
         return self.collect_params(select)
 
-    def set_nms(self, nms_thresh=0.3, nms_topk=400, post_nms=100):
+    def set_nms(self, nms_thresh, nms_topk, post_nms):
```
Comment (zhreshold) [inline gluoncv/model_zoo/rcnn/rcnn.py]: default values is ok here, it's different from network initialization

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py (hunk 1 above)
Comment (ijkguo) [reply gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: The loop is used for batch_size > 1. Each image goes through one nms with class dim as batch dim.

**Comment by zhreshold** [comment]
remaining concerns:

- add valid_thresh to box_nms
- revert to have default values for set_nms
- option 'use_crowd=False' in coco detection

Let's merge it in first

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py (hunk 2 above)
Comment (ijkguo) [reply gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: Setting a default thresh 0.0001. It is unlikely to change with respect to dataset or network so I hard code it.

File: gluoncv/data/mscoco/detection.py [hunk 5]
```diff
@@ -51,12 +53,13 @@ class COCODetection(VisionDataset):
 
     def __init__(self, root=os.path.join('~', '.mxnet', 'datasets', 'coco'),
                  splits=('instances_val2017',), transform=None, min_object_area=0,
-                 skip_empty=True):
+                 skip_empty=True, use_crowd=True):
```
Comment (zhreshold) [inline gluoncv/data/mscoco/detection.py]: use_crowd by default should be False?

**Comment by zhreshold** [comment]
Rebase to master to fix CI

File: gluoncv/data/mscoco/detection.py (hunk 5 above)
Comment (ijkguo) [reply gluoncv/data/mscoco/detection.py]: Not sure. Current 0.2 release did not filter crowd instances.

Comment (zhreshold) [reply gluoncv/data/mscoco/detection.py]: Okay, I think that should be fine, it depends on model as well.

**Comment by mli** [comment]
Job PR-206-12 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-206/12/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-206-13 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-206/13/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-206-14 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-206/14/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
The CI actually passed, but returned connection error. So I am merging this now.

**Pull request merged**